Media Library

vijaynandwani opened the pull request:
The media library requires gallery bundle. That's why this PR contains files from gallery bundle too.

vijaynandwani commented:
There were schema errors for dropzonejs. I'll open a issue and provide a patch.

![job 50 1 isthegeek media travis ci](https://cloud.githubusercontent.com/assets/2961631/16185984/5919e382-36e5-11e6-8ad9-380985396639.png)

I still have to modify the css and js files. But a basic design is ready for review. You can review it online too. Visit the media listing tab on http://winuall.cloudapp.net/drupal/node/add/media 
Username is admin and password is admin

Also I guess I still need to modify before_script.sh. It has some unnecessary directory traversal. I'll fix it soon.

vijaynandwani commented:
Please rerun the tests if they fail. Rerunning would fix the tests.

paranojik commented:
Why does the library require the gallery bundle?

vijaynandwani commented:
@paranojik the view has a filter to select the bundles. By default it shows media items for three bundles: image, video and gallery. That's why it requires the gallery bundle.

paranojik commented:
The view should not depend on the bundles. If a user creates a new bundle it should automatically appear as an option in the filter list.

So please, remove the gallery stuff from this PR.

vijaynandwani committed:
Media Library

slashrsm commented on config/install/core.entity_view_display.media.gallery.media_library.yml:
```diff
@@ -0,0 +1,35 @@
+    weight: 1
+    label: hidden
+    settings:
+      image_style: ''
```
I think that we should create an image style for this and use it across all bundles.

slashrsm commented on config/install/entity_browser.browser.media_library.yml:
```diff
@@ -0,0 +1,40 @@
+    735d146c-a4b2-4327-a057-d109e0905e05: 'Upload files'
+    f7742394-192d-4026-9947-055f7e167fb8: 'Files listing'
+  width: '650'
+  height: '500'
```
Could be bigger.

slashrsm commented on media.module:
```diff
@@ -22,3 +23,38 @@ function media_help($route_name, RouteMatchInterface $route_match) {
+ * Implements hook_form_alter().
+ */
+function media_form_alter(&$form, FormStateInterface &$form_state) {
+  if (isset($form['#form_id'])) {
```
Form ID should always be set. It is even 3rd argument to this function (see docs).

slashrsm commented on media.module:
```diff
@@ -22,3 +23,38 @@ function media_help($route_name, RouteMatchInterface $route_match) {
+  if (isset($form['#form_id'])) {
+    if ($form['#form_id'] == 'entity_browser_media_library_form') {
+      // Attach our library.
+      $form['#attached']['library'][] = 'media/view';
```
EB view is styled while user view is not. I think that you are attaching only to the first one here. We should make sure that styles are added consistently.

vijaynandwani replied:
I was working on the user view. I have attached it now. I'll update the cs and js files soon

paranojik commented on config/install/views.view.media_library.yml:
```diff
@@ -0,0 +1,646 @@
+        - 'config:core.entity_view_display.media.video.default'
+  page_1:
+    display_plugin: page
+    id: page_1
```
Wouldn't hurt giving it a more sensible machine name.

paranojik commented on config/install/views.view.media_library.yml:
```diff
@@ -0,0 +1,646 @@
+  page_1:
+    display_plugin: page
+    id: page_1
+    display_title: Page
```
...and title...

paranojik commented on css/media.view.css:
```diff
@@ -0,0 +1,205 @@
+    background: transparent;
+}
+
+.grid-item .views-field-entity-browser-select {
```
this doesn't sound right...

vijaynandwani replied:
I think this is right. You can check the html structure of the modal.

paranojik commented on js/media.view.js:
```diff
@@ -0,0 +1,49 @@
+      });
+
+      $('.grid-item').once('bind-click-event').click(function () {
+        var input = $(this).find('.views-field-entity-browser-select input');
```
more file browser references...

vijaynandwani replied:
This is not a file browser reference. views-field-entity-browser-select is the class of the select element of entities on the modal.

paranojik commented on media.module:
```diff
@@ -22,3 +24,46 @@ function media_help($route_name, RouteMatchInterface $route_match) {
+ */
+function media_views_pre_render(ViewExecutable $view) {
+  if (isset($view) && ($view->storage->id() == 'media_library')) {
+    $view->element['#attached']['library'][] = 'media/user';
```
Couldn't we use the same JS on all displays of the Media library view (Global library, entity browser and user library)?

vijaynandwani replied:
Okay I am doing that. File entity browser had different css files, that's why I made it like that.

vijaynandwani commented:
I need to make very small changes for global library. I'll squash the commits once they are done.

paranojik commented on config/install/core.entity_view_display.media.gallery.media_library.yml:
```diff
@@ -0,0 +1,36 @@
+    weight: 1
+    label: hidden
+    settings:
+      image_style: gallery_item
```
Shouldn't this be a "media_library_item" image style? We said gallery_item should be used in the display of the media gallery.

paranojik commented on config/install/core.entity_view_display.media.image.media_library.yml:
```diff
@@ -0,0 +1,35 @@
+    weight: 0
+    label: hidden
+    settings:
+      image_style: ''
```
style not set here?

paranojik commented on config/install/core.entity_view_display.media.video.media_library.yml:
```diff
@@ -0,0 +1,39 @@
+    weight: 1
+    label: hidden
+    settings:
+      image_style: ''
```
style not set here?

paranojik commented on config/install/image.style.media_library_item.yml:
```diff
@@ -0,0 +1,6 @@
+dependencies: {  }
+name: media_library_item
+label: 'Media library item'
+effects: {  }
```
We should size it to something like a thumbnail...

vijaynandwani replied:
It might interfere with the masonry layout. Just review it once without the thumbnail effect. If you still feel the need, I'll do it.

paranojik replied:
I just can't imagine the library to work (let alone be snappy) if you display original files (e.g. 10MB in size).

slashrsm replied:
We should definitely scale this images to a predictable size.

vijaynandwani replied:
Okay fixing it.

slashrsm commented on media.module:
```diff
@@ -22,3 +24,43 @@ function media_help($route_name, RouteMatchInterface $route_match) {
+function media_form_alter(&$form, FormStateInterface &$form_state) {
+  if ($form['#form_id'] == 'entity_browser_media_library_form') {
+    // Style the submit button.
+    $form['actions']['submit']['#value'] = t('Select Media Files');
```
This string is configurable on entity browser configuration now.

paranojik commented:
- The "Add media" link is missing on the library page.
- The global media library does not look the same as the user library
- Global media library should be just another page display in the "Media library" view, not a separate view.

vijaynandwani commented:
@paranojik 

"The global media library does not look the same as the user library"
You are talking about the media items information which is being displayed, right?
And should I change the access permission of the user library to "media: Edit own content" ?

paranojik commented:
@isthegeek Yes. The information displayed is different. I thought we agreed we'd manage that via view modes (not by adding fields in the view).
Yes. You can change the access permission to what you suggested.
Also, display the media add link on the library pages.

```
function media_menu_local_actions_alter(&$local_actions) {
    $local_actions['media.add']['appears_on'][] = 'view.media_library.user_media_library';
    $local_actions['media.add']['appears_on'][] = 'view.global_media_library.media_page_list';
}
```

...and don't forget to "merge" the global media library page into the media_library view. We shouldn't have to different views for this...

vijaynandwani commented:
I have fixed all the issues. Squashed the commits. I added couple of fields more in the view mode. I added the remaining fields in the view. Is it okay now?
